### ISSUE TITLE
Follow-up on GeoDataFrame._constructor refactoring (return type)

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#2133]
> Do you remember others? You also still have your other branches that did more extensive changes that are worth rebasing once this is merged?

Have had a look through my notes and the branches, here's what I've come up with:
* Special case of the first point, what should `df[['geometry2']]` return, if behaviour consistent with `__getitem__` is returned (specifically thinking of `.squeeze()` and making this consistent/ useful (although I am yet to find a situation where I've found this to be a problem)
* Implicit in what you've said in referencing the other branches - and in some other PRs, make `_geodataframe_constructor_with_fallback` smarter, to be aware of active geometry column. This will be needed to mimic `__getitem__` - I've already done some testing of this. What is probably more important however is if we can track the active geometry column in this way, it might be possible to clean up some of the special case handling of this in `apply`, `__finalize__`, `astype` and potentially other methods.

Points which are related to the resolution of inconsistency with "`__getitem__` behaviour" (which are probably secondary to what you are asking here, but I still think they're worth noting)
* If the "`__getitem__` behaviour" is opted for, does this have implications for `.rename`, and does that make `df.rename(columns={"geometry":"geom"})` returning a geodataframe with an invalid geometry inconsistent with the rest of the api? I'm not sure if there is a historical reason why `rename` doesn't also do `rename_geometry` under the covers, but that might be the logical direction if this is percieved as inconsistent (I remember this tripping me up when I was new to geopandas).
* A similar but less related tangent, what should happen on direct assignment to a geometry column it becomes invalid, e.g. `df['geometry'] =3`. I guess both of these could also be rephrased as an explicit question about api, "When should a GeoDataFrame return a DataFrame?" [I don't think either of these cases should return dataframes, but I think if return types are conditional, it would be nice if the logic for this was laid out specifically, rather than something one discovers by chance/ in confusion.

_Originally posted by @m-richards in https://github.com/geopandas/geopandas/issues/2060#issuecomment-927100352_

**Comment by m-richards** [comment]
Following the discussion on this at yesterdays meeting just wanted to recap that and try and articulate a direction to get this series of PR's merged (while I don't think only having some of them in should stop a 0.11 release, as we've already discovered there might be some edge cases and regressions to deal with and I think would be good to have them all part of a single release, rather than spreading that pain over two releases).

I also hope that having a clear plan might reduce some of the effort in reviewing these @jorisvandenbossche has been doing the heavy lifting there and I'd like to minimise further effort if possible.
1. Fix #2282. In the short term I think this will involve overloading `iloc` and the associated pandas class. (In the more longer term, I think it would be cleaner to move the `geoseries.geometry` attribute (which returns self) to `geoseries._geometry` to avoid this name collision. 
2. Resolve `__getitem__` behaviour in the presence of multiple geometry columns. This was left unresolved in #2060, with the view that any change to the indexing methods was an improvement. Given the discussion yesterday (along with the use case of using a GeoDataFrame as a container of Geoseries and operating on specific series, rather than the GeoDataFrame geometry attributes attached to the active geometry column, e.g in #2321). The change here would be that a return result with any columns of GeometryDtype would result in the return type being a GeoDataFrame (as opposed to currently, the presence of the active geometry column dictates the return type - this behaviour predates there being multiple geometry columns). The other aspect of this is if there needs to be a more formal way of denoting an "invalid" geodataframe - that might just mean updating the error message for the absence of the geometry column to be a bit more explicit.
3. Update the new behaviour from #2060 for the indexing methods, loc, iloc, drop, ... to be consistent with the behaviour in part 1. This could be the same PR but it would probably be easier to review separately.
4. Finish fixing #2296 - this is actually independent to the above two so I could have listed in first
5. Fix #2280 to be consistent with part 2 & 3 - this PR will need a bit of a rework since these two steps change the goal posts for it (I did wonder if it could make sense to do part 1 & 3 after this is merged (that would have been the original plan) but I can see a world where this PR takes a while to get right.
6. Clean up geodataframe.py, removing `__getitem__`, and hopefully some conditional metadata logic in some other methods. `astype` and `convert_dtypes` should be close to completely removable.

Well this is longer than I would have liked but hopefully it gives some clarity, happy to hear other thoughts as well.

**Comment by m-richards** [comment]
An initial draft of part 2: #2329

**Comment by jorisvandenbossche** [comment]
(note, @m-richards, I started to write this before your post appeared, so posting this comment anyway, will look at your comment afterwards)

> Decide what to do with results that loose the "geometry" column, but still have another geometry dtype column (currently this is a bit inconsistent, `__getitem__` gives DataFrame, but most other methods result in an "invalid" GeoDataFrame with `_geometry_column_name` not set)

This first point of the list above is something we touched upon in the community meeting yesterday, and so something we should try to resolve / decide upon for the 0.11 release. 
It is also related to the recent issue that was brought up in https://github.com/geopandas/geopandas/issues/2321

So let's maybe first show the current situation, with an example of a GeoDataFrame with a geometry column but which is not the "active" one:

```python
df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
df["geom2"] = df.geometry.centroid
df2 = df.loc[:, ["name", "geom2"]]
```
```
>>> type(df2)
geopandas.geodataframe.GeoDataFrame
>>> df2._geometry_column_name
'geometry'
>>> df2.area
...
AttributeError: No geometry data set yet (expected in column 'geometry'.)
```

And you get the same result with the `GeoDataFrame` constructor if passing data without "geometry" column:

```
df3 = geopandas.GeoDataFrame({"col": [1, 2, 3], "geom": geopandas.points_from_xy([1, 2, 3], [1, 2, 3])})

>>> type(df3)
geopandas.geodataframe.GeoDataFrame
>>> df3._geometry_column_name
'geometry'
df3.area
...
AttributeError: No geometry data set yet (expected in column 'geometry'.)
```

One good reason why it is useful that those still return a GeoDataFrame, even though "invalid", is that this ensure that if you access the other geometry-dtyped column, you still get a GeoSeries with all direct access to the geospatial functionality (this was also brought up in https://github.com/geopandas/geopandas/issues/2321):

```
>>> type(df3["geom"])
geopandas.geoseries.GeoSeries
```

If we would no longer return a GeoDataFrame but a DataFrame instead, the above would return a plain pd.Series.

Following what I mentioned yesterday, we could maybe also look into improving the experience of such "invalid" GeoDataFrames (i.e. where the "active geometry column" is not present):

* Should we maybe keep `_geometry_column_name` unset (`None`)? (I don't know if this actually would help in practice, it might just make it a bit easier to check if the active column is set or not)

* We could improve the error message you get when doing a geospatial operation on a GeoDataFrame with no active geometry column set. The current "AttributeError: No geometry data set yet (expected in column 'geometry'.)" is actually already quite clear in that it expects to find a "geometry" column, but I think we can make this even more helpful with a message like:

  > You are calling a geospatial method on the GeoDataFrame, but the default active geometry column to use has not been set. There are columns with geometry data type ([...]), and you can either set this as the active geometry with `df.set_geometry("name")` or access this column as a GeoSeries (df["name"]) and call the method directly on that. 

* Fix the `.crs` situation where this can still return something even if the active geometry column is not there

**Comment by jorisvandenbossche** [comment]
> 2\. .... The change here would be that a return result with any columns of GeometryDtype would result in the return type being a GeoDataFrame (as opposed to currently, the presence of the active geometry column dictates the return type - this behaviour predates there being multiple geometry columns). The other aspect of this is if there needs to be a more formal way of denoting an "invalid" geodataframe - that might just mean updating the error message for the absence of the geometry column to be a bit more explicit.

Yes, I think that more or less matches with what I also recollected in my comment above. 

> 6\. Clean up geodataframe.py, removing `__getitem__`, and hopefully some conditional metadata logic in some other methods. `astype` and `convert_dtypes` should be close to completely removable.

It would be nice if we could remove `__getitem__`, and from a quick look it seems that should indeed be possible. In any case we should otherwise update it to match the behaviour outlined in point 2 (i.e. also return GeoDataFrame if any geometry-dtype column is present)

**Comment by martinfleis** [comment]
> Fix the .crs situation where this can still return something even if the active geometry column is not there (DEP: Assigning CRS to a GeoDataFrame without a geometry column #2268 is first step, but in addition we can maybe also explicitly deprecate the ability to access .crs is the active geometry column is not present).

I think that the logic being now enforced in #2268 says that `.crs` can work only as an alias `.geometry.crs` and accessing `.crs` with no active geometry should raise the same warning as accessing `.area` for example. But you are probably right that there may be cases where people rely on this and we should deprecate that before removing it fully. But that deprecation should surely go to 0.11.

**Comment by jorisvandenbossche** [comment]
https://github.com/geopandas/geopandas/pull/2268 only enforces that you can't manually set the crs on a GeoDataFrame with no active geometry or create such a GeoDataFrame, but you can still end up with such a GeoDataFrame starting from a valid one. For example:

```python
>>> df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_lowres"))
# or any other operation that results in a GeoDataFrame that lost the connection with the active geometry column
>>> df2 = df.rename(columns={"geometry": "geoms"})
# this still works
>> df2.crs
<Geographic 2D CRS: EPSG:4326>
Name: WGS 84
...
# while accessing the geometry doesn't
>>> df2.geometry
...
AttributeError: No geometry data set yet (expected in column 'geometry'.)
```

It is quite similar to the case as you fixed in osmnx (but then for cases where we would preserve the GeoDataFrame class, while in the osmnx case it already failed now because it started to return a DataFrame)

**Comment by m-richards** [comment]
>Should we maybe keep _geometry_column_name unset (None)? (I don't know if this actually would help in practice, it might just make it a bit easier to check if the active column is set or not)

I think we should definitely do this when the active geometry column is lost but there are other geometry columns. It's nice and explicit, and should help point out the cases where this is handled properly (it is set to None) from the cases where it is not (it's still set to a column label that's no longer present).

**Comment by m-richards** [comment]
Now that #2329 is merged, I'll have a look point 3 in my list above and work out if it's best to update #2280 or do this in yet another PR.
